spec.md

```diff
@@ -975,6 +1173,14 @@ message ValidateVolumeCapabilitiesRequest {
   // Attributes of the volume to check. This field is OPTIONAL and MUST
   // match the attributes of the Volume identified by `volume_id`.
   map<string, string> volume_attributes = 3;
+
+  // Specifies where (regions, zones, racks, etc.) the caller believes
+  // the volume is accessible from.
+  // An caller may specify multiple topologies to indicates they believe
+  // the volume to be accessible from multiple locations.
+  // This field is OPTIONAL. This field MAY only be set if the plugin
+  // advertises the ACCESSIBILITY_CONSTRAINTS capability.
+  repeated Topology accessible_topology = 4;
```

Comment:
This does not seem necessary given that `Volume` already has this information.

Reply:
I don't understand: `Volume` isn't part of this RPC request

Reply:
Volume is not part of `ValidateVolumeCapabilitiesRequest` (maybe it should be, but that is beyond the scope here), and it would be nice for a CO to be able to validate its assumptions about accessibility of the volume.

Reply:
In Kubernetes we allow the user to directly specify a pre-provisioned volume and all it's attributes (including topology information). So basically, step 1 can be skipped. In that case, it would be nice to validate assumptions about accessibility too.

Reply:
+1 on the `capacity_bytes`. Honestly we should just pass the full volume object in addition to the `VolumeCapabilities` but that is out of scope for this PR so I will not do that here.

> can we put topology information in VolumeCapability?

I'm ok with that. @jdef?

Reply:
@saad-ali, i had a chat with @jdef and @chhsia0 on this. The reason I was questioning this choice initially was due to consistency (I believe that if something is not consistent, usually something is missing or incorrect).

I think the reason we originally introduced `ValidateVolumeCapabilities`, instead of putting `VolumeCapability` in `Volume` object is because it's very hard to enumerate all possible combinations of supported volume capabilities. Instead, we let the CO to drive the volume capability test using this call.

Ideally, the arguments in this call should be consistent with those in `CreateVolumeRequest` (i.e., `CapacityRanage` and `TopologyRequirement`) because that's what COs should care about when dealing with pre-existing volumes the same way as if they're dealing with dynamically provisioned volumes. To map that to the k8s world, that means this call probably needs to be called when binding a PVC to a PV, rather than validating if a PV is valid or not. But I guess this is a much bigger change?

I think moving topology into volume capability is probably too big of a change (which I haven't thought through). At this stage, I think I am fine keeping as it is right now, with the plan to add `capacity_bytes` to `ValidateVolumeCapabilities`. Those fields need to be optional, if not specified, the SP won't check those against the volume.

Reply:
> Ideally, the arguments in this call should be consistent with those in CreateVolumeRequest (i.e., CapacityRanage and TopologyRequirement) because that's what COs should care about when dealing with pre-existing volumes the same way as if they're dealing with dynamically provisioned volumes. To map that to the k8s world, that means this call probably needs to be called when binding a PVC to a PV, rather than validating if a PV is valid or not. But I guess this is a much bigger change?

Agreed. We should consider it pre-1.0 though. Opened https://github.com/container-storage-interface/spec/issues/242 to track that.

> I think moving topology into volume capability is probably too big of a change (which I haven't thought through). At this stage, I think I am fine keeping as it is right now, with the plan to add capacity_bytes to ValidateVolumeCapabilities. Those fields need to be optional, if not specified, the SP won't check those against the volume.

SGTM, will leave as is for now.